### PR TITLE
[DataLoader] Fix DataLoader to support non-Tensor batch data

### LINKED ISSUE #77754: paddle的Dataset, DataLoader不支持返回非Tensor类数据
### bug描述 Describe the Bug

如题，如果是使用torch框架运行如下代码：
```Python
#from paddle.io import Dataset, DataLoader
from torch.utils.data import DataLoader, Dataset
# define a random dataset
class RandomDataset(Dataset):
    def __init__(self, num_samples):
        self.num_samples = num_samples

    def __getitem__(self, idx):
        return "a"

    def __len__(self):
        return self.num_samples

def f(d):
    return d

dataset = RandomDataset(20)

loader = DataLoader(dataset,
                    batch_size=10,
                    shuffle=True,
                    drop_last=True,
                    collate_fn=f)

for e in range(1):
    for i, j in enumerate(loader):
        print(j)

```
结果为
```Text
E:\code\paddleaudio补全\别的贡献计划\ECD_but_paddle>python C:\Users\15895\Desktop\functions.py
['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a']
['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a']

E:\code\paddleaudio补全\别的贡献计划\ECD_but_paddle>
```
如果是把运行框架换成paddle:
```Python
from paddle.io import Dataset, DataLoader
#from torch.utils.data import DataLoader, Dataset
# define a random dataset
class RandomDataset(Dataset):
    def __init__(self, num_samples):
        self.num_samples = num_samples

    def __getitem__(self, idx):
        return "a"

    def __len__(self):
        return self.num_samples

def f(d):
    return d

dataset = RandomDataset(20)

loader = DataLoader(dataset,
                    batch_size=10,
                    shuffle=True,
                    drop_last=True,
                    collate_fn=f)

for e in range(1):
    for i, j in enumerate(loader):
        print(j)

```
结果为
```Text
E:\code\paddleaudio补全\别的贡献计划\ECD_but_paddle>python C:\Users\15895\Desktop\functions.py

E:\code\paddleaudio补全\别的贡献计划\ECD_but_paddle>
```
经过本人两个小时的debug，触发迭代终止的逻辑似乎写在C++里面了，大概是C++那边判断类型不是Tensor就给直接抛出异常了。paddle框架的DataLoader应该支持非Tensor类型数据的返回，否则将会影响众多项目的复现。

### 其他补充信息 Additional Supplementary Information

首当其冲的就是 https://github.com/PFCCLab/paddle_geometric 这个库
因为这个库的加载数据的逻辑是，返回一个包含自定义数据类型的列表
我滴妈呀，反正我用paddle_geometric的时候是不敢用框架内的加载逻辑了
感觉当下想要解决这个问题，最好的办法就是自己写一个DataLoader类，难绷
可惜了我又因此debug了两个小时，复现模型时时间居然主要浪费在和框架斗智斗勇这一块
总之， #77752 和这个issue如果不能想办法解决，paddle_geometric这个库就别想活了（
毕竟人家那个库的逻辑在那，要么修复框架，要么对库逻辑重构，确实没办法

### PR DESCRIPTION
**Pull request by ShigureNyako**
### PR Category
Execute Infrastructure

### PR Types
Bug fixes

#### What PR does
This PR fixes DataLoader / PyReader EOF detection so that non-Tensor batch data is supported correctly. Previously, an empty batch could be treated as EOF directly, which caused the reading process to stop early even when the batch itself was valid.

It also fixes the Python-side DataLoader iterator logic so that a real EOF (empty tensor list with no pending recorded structure) raises `StopIteration` instead of trying to restore a non-existent batch structure.

#### Why need this PR
In the current implementation, `ins.empty()` is used to determine whether the reader has reached EOF. However, for non-Tensor data paths, a valid empty batch can still be produced, for example an empty `TensorArray`. Treating every empty batch as EOF is incorrect and breaks normal data loading behavior.

After refining the C++ EOF semantics, the Python iterator still assumed every empty tensor list corresponded to a recorded structure entry. On the final EOF path for pure non-Tensor batches, this led to `self._structure_infos.pop(0)` raising `IndexError: pop from empty list`.

#### How to fix
This PR refines the EOF semantics in both the C++ reader path and the Python DataLoader iterator:
- Add a `HasReachedEnd()` virtual interface to `ReaderBase` to distinguish real EOF from an empty batch.
- Let `PyReader` maintain its own `reached_end_` state instead of binding EOF to empty batch data.
- Update `read_op` to check EOF with `ins.empty() && reader->HasReachedEnd()`.
- Expose the related interface through pybind and add tests for the empty-batch-but-not-EOF case.
- In Python single-process and multi-process DataLoader iterators, treat `len(data) == 0 && len(self._structure_infos) == 0` as real EOF and raise `StopIteration` before restoring batch structure.

#### Test performed
- Added C++ unit test: `test/cpp/fluid/reader/py_reader_test.cc`
- Updated `test/cpp/fluid/reader/CMakeLists.txt` to include the new test
- Added Python integration tests for non-Tensor batch support: `test/legacy_test/test_multiprocess_dataloader_dynamic.py`
- Local syntax check: `python3 -m py_compile python/paddle/io/dataloader/dataloader_iter.py test/legacy_test/test_multiprocess_dataloader_dynamic.py`

#### Related issue
- Fixes #77754

#### Notes
- This PR supersedes #78213, which was based on a branch containing unrelated commits.

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/io/dataloader/dataloader_iter.py [hunk 1]
```diff
@@ -306,6 +306,8 @@ def __next__(self):
                 data = core.eager.read_next_tensor_list(
                     self._reader.read_next_list()[0]
                 )
+                if len(data) == 0 and len(self._structure_infos) == 0:
+                    raise StopIteration
```
Comment (zyfncg) [inline python/paddle/io/dataloader/dataloader_iter.py:310]: 这里修改 _structure_infos 的目的是什么，和前面 C++ 层的修改有什么联系
